_posts/2023-07-10-introducing-extensions-for-opensearch.md

```diff
@@ -0,0 +1,127 @@
+
+We were also able to further decrease the number of data nodes in order to optimize searching data and indexing results. For example, because most data nodes in the original performance benchmark had insignificant usage other than indexing results, we tested a cluster with only 24 c5.2xlarge data nodes along with 1x r5.16xlarge extension node, a further 33% reduction. We believe that with further experimentation and fine-tuning, you can find an even more optimal mix, depending on other resource requirements for the cluster.
+
+### 2. Increase indexing throughput
+
+While we have not run this experiment, we believe you can further independently fine-tune the cluster by increasing sharding and achieve better throughput in indexing AD results.
+
+### 3. Increase the rate of generating historical results
+
+AD’s historical HCAD restricts the rate of historical time period results, limits the use of AD to half the heap, and sets the maximum for detection at 10 simultaneous threads to protect the cluster from being maxed out. You can raise these limits by adding more CPUs or RAM, swapping space with SSDs independently from the rest of the cluster, or horizontally scaling extension nodes.
+
+## Conclusion
+
+Extensions save costs and improve performance by right-sizing all resources to the task being performed. In our test, we did not need memory-optimized servers for the whole cluster to perform historical AD and therefore could achieve a 33% cost reduction per data node.
+
+As a plugin, AD shares CPU and memory with OpenSearch I/O and must limit its own resource consumption in order to ensure the cluster’s stability. For example, in the AD plugin HCAD tests, models were limited to using half of the heap, which itself was half of the node’s memory. In the AD extension scenario, it was possible to scale the heap to consume most of the server’s memory and fully utilize 64 vCPUs dedicated only to AD tasks.
+
+Other existing plugins that could similarly benefit from this model include all plugins that have memory- or CPU-bound concerns, such as ML Commons, k-NN, and Reporting. This new technology also allows implementing new resource-heavy extensions, such as reranking, data transformation, big data processing, or video indexing.
+
+## Next steps
+
+We are thrilled with the progress we have made with extensibility so far and the early performance benchmarking results. Before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading the [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting on the [OpenSearch forum](https://forum.opensearch.org/) or creating an issue in the [`opensearch-sdk-java` repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!
```

Comment:
We should add transparency here on the level of investment. We are excited to gain feedback, but need to be clear we don't have a timeline to make this generally available. 
**Recommendation** put this : 

We are thrilled with the progress we have made with extensibility so far and the early performance benchmarking results. Moving forward, in the short term we will shift our focus to using the learnings from the Extensibility project to enable extensibility in our OpenSearch Machine Learning offerings. The goal is to deliver a spec-based application framework where users will be able to create no-code ML-powered applications in OpenSearch, and ultimately deliver a simple one-click way for users to test, build and deploy ML-powered applications using OpenSearch. Additionally, before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading the [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting on the [OpenSearch forum](https://forum.opensearch.org/) or creating an issue in the [`opensearch-sdk-java` repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!

Reply:
Edited version: We are thrilled with the progress we have made with extensibility so far and with the early performance benchmarking results. Moving forward, in the short term we will shift our focus to using the learnings from the Extensibility project to enable extensibility in our OpenSearch Machine Learning (ML) offerings. The goal is to deliver a spec-based application framework where users will be able to create no-code ML-powered applications in OpenSearch and ultimately deliver a simple one-click way for users to test, build, and deploy ML-powered applications using OpenSearch. Additionally, before we commit to releasing this experimental feature and the SDK as generally available, we would like your feedback. After reading the [extension documentation](https://opensearch-project.github.io/opensearch-sdk-java/), try implementing an experimental extension and give us feedback by posting on the [OpenSearch forum](https://forum.opensearch.org/) or creating an issue in the [opensearch-sdk-java repo](https://github.com/opensearch-project/opensearch-sdk-java/). We can’t wait to hear from you!